### PR TITLE
protocols: add readme

### LINKED ISSUE #52: Update ./protocols package documentation / readme
`./protocols` currently has no readme.

New readme should:
- [ ] describe the interfaces defined in the package
- [ ] contain a road map for implementation status on different protocols
- [ ] Address [this comment](https://github.com/statechannels/go-nitro/pull/42#discussion_r749465761).
- [ ] ... ?

Old readme contained context on different implementation strategies (state-chart vs flowChart).

- [ ] Package documentation should include some discoverable reference to these thoughts.

### PR DESCRIPTION
**Pull request by geoknee**
Closes #52 and #32

This PR adds a readme ([preview here!](https://github.com/statechannels/go-nitro/tree/protocols-readme/protocols))

The readme helps the reader to understand what an objective is, how it relates to a protocol, and also has a roadmap showing which protocols have been implemented so far.

### CODE REVIEW

File: protocols/readme.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+The imperative shell will typically be responsible for:
+
+- spawning new objective instances of one type or another
+- reading and persisting objectives to and from a store.
+- "updating" objectives by passing events to their `Update(event)` function (an updated copy is returned)
+- "cranking" objectives by calling `Crank()` (side effects are returned)
+- executing side effects
+
+There is an architectural decision record contained in a hidden folder in this directory. [Link](./.adr/readme.md)
```
Comment (NiloCK) [inline protocols/readme.md:29]:
```suggestion
This folder contains a hidden directory with architectural decision records. [Link](./adr)
```

Plus a rename of `.adr/readme.md` to something like `0000-flowchart-vs-statechart.md`.

**Review by NiloCK — APPROVED** [review]
:1st_place_medal:

**Comment by geoknee** [comment]
Thanks @NiloCK . I'm asking @andrewgordstewart for a second review because:
- he raised #32 in a previous PR review and I'd like to know if he thinks we have addressed it
- the adr hidden folder feels like a reasonably significant addition to our processes, so some more feedback seems sensible

**Comment by geoknee** [comment]
This PR to be split into:
* a readme-only PR (Closing #52 and #32), for a reasonably quick review from @andrewgordstewart 
* the addition of the .adr folder (this is less urgent and could wait a week, at least)

**Comment by NiloCK** [comment]
My understanding Re PR Splitting:

The (an) approach is to take this branch, which does NOT run afoul of this rule:

> Do not rebase commits that exist outside your repository and that people may have based work on.

... and rewrite its history into two branches. The current `protocols-readme` branch can retain the changes wrt readme.md, and another branch (which will become another PR) contains the `./adr` changes.

**Comment by andrewgordstewart** [comment]
> The (an) approach is to take this branch, which does NOT run afoul of this rule:

This is exactly the approach I would suggest.

**Comment by NiloCK** [comment]
**Plan** (documenting ideas before they do not work out):
- `git reset` the current branch back to `main`
- re-commit protocols/readme work
- push (force push?)
At this point, THIS pr is as we want it. Contains the new readme and nothing else.
- stash remaining changes
- `checkout -b adr` from `main`
- stash pop adr change set and commit it
- push + make second PR

**Comment by andrewgordstewart** [comment]
> **Plan** (documenting ideas before they do not work out):
> 
> * `git reset` the current branch back to `main`
> * re-commit protocols/readme work
> * push (force push?)
>   At this point, THIS pr is as we want it. Contains the new readme and nothing else.
> * stash remaining changes
> * `checkout -b adr` from `main`
> * stash pop adr change set and commit it
> * push + make second PR

Here's my flow in this situation:

1. checkout `protocols-readme` branch
2. create a new branch `adr` from `HEAD` (Now you have two copies of the complete work on two different branches)
3. reset HEAD to `main`, check in `adr`-related code, discard the rest, push
4. checkout `protocols-readme`, reset `HEAD` to `main`, check in `readme`-related code, discard the rest, push

If you don't want to force-push, make a third copy `colin-readme` from the `protocols-readme` branch.

File: protocols/readme.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+To implement the protocols, we make use of an abstraction we call the `Objective`. Objectives are named for the goal which the off-chain protocol aspires to achieve.
+
+The following table shows some example objectives and their implemetation status:
+
+| Objective type      | Implemented |
+| ------------------- | ----------- |
+| `direct-funding`    | x           |
+| `direct-defunding`  |             |
+| `virtual-funding`   |             |
+| `virtual-defunding` |             |
```
Comment (andrewgordstewart) [inline protocols/readme.md:14]: nit: It almost sounds like `directly-fund` is a better name for `Objective`. hard to imagine this is worth the refactor though?

**Review by andrewgordstewart — APPROVED** [review]
This is great.

The architecture described here may deserve its own ADR 😆 (In all seriousness, we should ask for PolyCrypt's feedback on this architecture.)

File: protocols/readme.md (hunk 2 above)
Comment (geoknee) [reply protocols/readme.md:14]: I agree that's a better name. Let's think about if and when to make the change?

Comment (andrewgordstewart) [reply protocols/readme.md:14]: yep, this is not a blocker at  all

**Pull request merged**